File: Patrones Estructurales/Adapter/adapter.js
// 1. Target (Lo que el cliente espera usar)
class EnchufeAmericano {
    conectar110V() {
        throw new Error("Este método debe ser implementado");
    }
}

// 2. Adaptee (La clase incompatible que ya tenemos)
class LicuadoraEuropea {
    conectar220V() {
        console.log("Licuadora funcionando a 220V.");
    }
}

// 3. Adapter (El intermediario)
class AdaptadorEuropeo extends EnchufeAmericano {
    constructor(licuadora) {
        super();
        this.licuadora = licuadora;
    }

    conectar110V() {
        console.log(">>> Adaptando corriente de 110V a 220V...");
        this.licuadora.conectar220V();
    }
}

// Uso
const miLicuadora = new LicuadoraEuropea();
const adaptador = new AdaptadorEuropeo(miLicuadora);

adaptador.conectar110V(); // El cliente usa la interfaz americana

File: Patrones Estructurales/Adapter/adapter.py
# 1. Target (Interfaz esperada)
class EnchufeAmericano:
    def conectar_110v(self):
        pass

# 2. Adaptee (Clase incompatible)
class LicuadoraEuropea:
    def conectar_220v(self):
        print("Licuadora funcionando a 220V.")

# 3. Adapter (Intermediario)
class AdaptadorEuropeo(EnchufeAmericano):
    def __init__(self, licuadora):
        self.licuadora = licuadora

    def conectar_110v(self):
        print(">>> Adaptando corriente de 110V a 220V...")
        self.licuadora.conectar_220v()

# Uso
if __name__ == "__main__":
    mi_licuadora = LicuadoraEuropea()
    adaptador = AdaptadorEuropeo(mi_licuadora)
    
    adaptador.conectar_110v()